doc/monads/applicatives.lean

```diff
@@ -0,0 +1,329 @@
+/-!
+# Applicative Functors
+
+Building on [Functors](functors.lean.md) is the [Applicative
+Functor](https://en.wikipedia.org/wiki/Applicative_functor). For simplicity, you can refer to these
+simply as "Applicatives". These are a little tricker than functors, but still simpler than monads.
+Let's see how they work!
+
+## What is an Applicative Functor?
+
+An applicative functor is an defines a default or "base" construction for an object and allows
+function application to be chained across multiple instances of the structure. All applicative
+functors are functors, meaning they must also support the "map" operation.
+
+## How are Applicatives represented in Lean?
+
+An [applicative functor](https://en.wikipedia.org/wiki/Applicative_functor) is an intermediate
+structure between `Functor` and `Monad`. It mainly consists of two operations:
+
+* `pure : α → F α`
+* `seq : F (α → β) → F α → F β` (written as `<*>`)
+
+The `pure` operator specifies how you can wrap a normal object `α` into an instance of this structure `F α`.
+This is the "default" mechanism mentioned above.
+
+The `seq` operator gives a notion of evaluation order to the effects, where the first argument is
+executed before the second, but unlike a monad the results of earlier computations cannot be used to
+define later actions.
+
+Applicative in Lean is built on some helper type classes, `Functor`, `Pure` and `Seq`:
+
+```lean,ignore
+class Applicative (f : Type u → Type v) extends Functor f, Pure f, Seq f, SeqLeft f, SeqRight f where
+```
+
+Notice that as with `Functor` it is also a type transformer `(f : Type u → Type v)` and notice the
+`extends Functor f` is ensuring the base Functor also performs that same type transformation.
+
+As stated above, all applicatives are then functors. This means you can assume that `map` already
+exists for all these types.
+
+The `Pure` base type class is a very simple type class that supplies the `pure` function.
+
+```lean,ignore
+class Pure (f : Type u → Type v) where
+   pure {α : Type u} : α → f α
+```
+
+You can think of it as lifing the result of a pure value to some monadic type. The simplest example
+of `pure` is the `Option` type:
+
+-/
+#eval (pure 10 : Option Nat)  -- some 10
+/-!
+
+Here we used the `Option` implementation of `pure` to wrap the `Nat 10` value in an `Option Nat`
+type resulting in the value `some 10`, and in fact if you look at the Monad instance of `Option` , you
+will see that `pure` is indeed implemented using `Option.some`:
+
+-/
+instance : Monad Option where
+    pure := Option.some
+/-!
+
+The `Seq` type class is also a simple type class that provides the `seq` operator which can
+also be written using the special syntax `<*>`.
+
+```lean,ignore
+class Seq (f : Type u → Type v) : Type (max (u+1) v) where
+  seq : {α β : Type u} → f (α → β) → (Unit → f α) → f β
+```
+
+The `seq` operator allows you to chain operations by wrapping a function in a structure. The name
```

Comment:
There are two sentences starting with "The `seq` operator ...", but no connection between them.

Reply:
We should also explain why we have `Unit -> ...` here.

Reply:
Perhaps, you could add a link to the Lazy Evaluation section, or just mention you will discuss it later.

Reply:
I carefully crafted the doc to avoid that complication until the Lazy Evaluation section because I don't want to side track the reader from what is important about seq, this is why I jump to using the `<*>` notation instead of showing how to call `seq` directly.  I think a forward reference would only complicate things unnecessarily.

Reply:
It looks confusing to me. There are two `seq`s in the same section with different types and no explanation why.